### PR TITLE
grpc-json: use gRPC status to HTTP status code mapping

### LINKED ISSUE #2006: Return appropriate HTTP error code when using gRPC-JSON Transcoder
**Issue Template**

*Title*: *Return appropriate HTTP error code when using gRPC-JSON Transcoder*

*Description*:

I have a .proto
```
syntax = "proto3";

package helloworld;

import "annotations.proto";

// The greeting service definition.
service Greeter {
  // Sends a greeting
  rpc SayHello (HelloRequest) returns (HelloReply) {
    option (google.api.http) = {
      post: "/api/v1/hello"
      body: "*"
    };
  }
}

// The request message containing the user's name.
message HelloRequest {
  string name = 1;
}

// The response message containing the greetings
message HelloReply {
  string message = 1;
}
```

And a Node.js server
```
var PROTO_PATH = './hello.proto';

var grpc = require('grpc');
var hello_proto = grpc.load(PROTO_PATH).helloworld;

function sayHello(call, callback) {
    if (call.request.name === "Error") {
        var err = new Error('PermissionDenied');
        err.code = grpc.status.PERMISSION_DENIED;
        return callback(err)
    }
    
    return callback(null, {message: 'Hello ' + call.request.name});
}

function main() {
    var server = new grpc.Server();
    server.addService(hello_proto.Greeter.service, {sayHello: sayHello});
    server.bind('0.0.0.0:50051', grpc.ServerCredentials.createInsecure());
    server.start();
}

main();
```

Then I run Envoy with following config
```
{
    "listeners": [{
        "address": "tcp://0.0.0.0:80",
        "filters": [{
            "type": "read",
            "name": "http_connection_manager",
            "config": {
                "server_name": "nextapi",
                "codec_type": "auto",
                "stat_prefix": "ingress_http",
                "use_remote_address": true,
                "access_log": [{"path": "/tmp/envoy.access.log"}],
                "route_config": {
                    "virtual_hosts": [{
                        "name": "grpc_service",
                        "domains": ["*"],
                        "routes": [{
                            "prefix": "/",
                            "cluster": "grpc1"
                        }]
                    }]
                },
                "filters": [
                    {
                        "type": "both",
                        "name": "grpc_json_transcoder",
                        "config": {
                            "proto_descriptor": "/etc/hello.pb",
                            "services": ["helloworld.Greeter"],
                            "print_options": {
                                "add_whitespace": false,
                                "always_print_primitive_fields": false,
                                "always_print_enums_as_ints": false,
                                "preserve_proto_field_names": false
                            }
                        }
                    },    
                    {
                        "type": "decoder",
                        "name": "router",
                        "config": {}
                    }
                ]
            }
        }]
    }],
    "admin": {
        "access_log_path": "/tmp/admin_access.log",
        "address": "tcp://0.0.0.0:9901"
    },
    "cluster_manager": {
        "clusters": [{
            "name": "grpc1",
            "connect_timeout_ms": 250,
            "type": "static",
            "lb_type": "random",
            "features": "http2",
            "hosts": [
                {"url": "tcp://192.168.99.100:50051"} 
            ]
        }]
    }
}
```

Test with curl
```
curl -X POST http://192.168.99.100/api/v1/hello -d '{"name":"Jim"}' -v
*   Trying 192.168.99.100...
* Connected to 192.168.99.100 (192.168.99.100) port 80 (#0)
> POST /api/v1/hello HTTP/1.1
> Host: 192.168.99.100
> User-Agent: curl/7.43.0
> Accept: */*
> Content-Length: 14
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 14 out of 14 bytes
< HTTP/1.1 200 OK
< content-type: application/json
< grpc-accept-encoding: identity,deflate,gzip
< accept-encoding: identity,gzip
< x-envoy-upstream-service-time: 9
< grpc-status: 0
< grpc-message: OK
< content-length: 23
< date: Tue, 07 Nov 2017 15:12:54 GMT
< server: envoy
< 
* Connection #0 to host 192.168.99.100 left intact
{"message":"Hello Jim"}
```

Then with error case `("name":"Error")`
```
curl -X POST http://192.168.99.100/api/v1/hello -d '{"name":"Error"}' -v
*   Trying 192.168.99.100...
* Connected to 192.168.99.100 (192.168.99.100) port 80 (#0)
> POST /api/v1/hello HTTP/1.1
> Host: 192.168.99.100
> User-Agent: curl/7.43.0
> Accept: */*
> Content-Length: 16
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 16 out of 16 bytes
< HTTP/1.1 200 OK
< content-type: application/json
< grpc-status: 7
< grpc-message: PermissionDenied
< x-envoy-upstream-service-time: 7
< date: Tue, 07 Nov 2017 15:12:52 GMT
< server: envoy
< content-length: 0
< 
* Connection #0 to host 192.168.99.100 left intact
```

gRPC server returning error with status code 7 (PERMISSION_DENIED) but HTTP status code is 200 (OK). I expect HTTP status code should be 403 (Forbidden). I refer to https://cloud.google.com/apis/design/errors#error_codes. 

Did I missed something in the config or currently it's not available in envoy? How error code in gRPC can be translated to appropriate HTTP error code with gRPC-JSON transcoder?

### PR DESCRIPTION
**Pull request by lizan**
*Description*:
Previously the while the `grpc-status` is propagated, the HTTP status code was blindly set to 200, this PR change it to use gRPC status to HTTP status code mapping.

Fixes #2006 

*Risk Level*: Low

*Testing*:
a integration test covers this.

*Release Notes*:
Note the gRRPC JSON transcoder filter behavior is now return HTTP status mapped from grpc-status.

### CODE REVIEW

File: source/common/grpc/json_transcoder_filter.cc [hunk 1]
```diff
@@ -338,13 +347,12 @@ Http::FilterTrailersStatus JsonTranscoderFilter::encodeTrailers(Http::HeaderMap&
...
   }
 
-  const Http::HeaderEntry* grpc_status_header = trailers.GrpcStatus();
-  if (grpc_status_header) {
-    uint64_t grpc_status_code;
-    if (!StringUtil::atoul(grpc_status_header->value().c_str(), grpc_status_code)) {
-      response_headers_->Status()->value(enumToInt(Http::Code::ServiceUnavailable));
-    }
-    response_headers_->insertGrpcStatus().value(*grpc_status_header);
+  Optional<Status::GrpcStatus> grpc_status = Common::getGrpcStatus(trailers);
```
Comment (mattklein123) [inline source/common/grpc/json_transcoder_filter.cc]: nit: const

**Comment by dhawani** [comment]
@lizan Does this support custom error responses ?

**Comment by lizan** [comment]
@dhawani If you mean custom `grpc-message` then yes you should still see that in response header. If not, can you elaborate more what is your custom error response?

**Comment by elvizlai** [comment]
@lizan 
We have defined some code with our need. For example: 10000.

But It look like when envoy grpc transcoder meed code greater than `Status::GrpcStatus::MaximumValid` will return http status `ServiceUnavailable`.

Can this be changed or configable?

**Pull request merged**